utils/zerovec/src/zerovec/mod.rs

```diff
@@ -0,0 +1,500 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(feature = "serde")]
+mod serde;
+
+use crate::ule::*;
+use std::fmt;
+
+/// A zero-copy vector for fixed-width types.
+///
+/// `ZeroVec<T>` is designed as a drop-in replacement for `Vec<T>` in situations where it is
+/// desirable to borrow data from an unaligned byte slice, such as zero-copy deserialization.
+///
+/// `T` must implement [`AsULE`], which is auto-implemented for a number of built-in types,
+/// including all fixed-width multibyte integers.
+///
+/// # How it Works
+///
+/// `ZeroVec<T>` represents a slice of `T` as a slice of `T::ULE`. The difference between `T` and
+/// `T::ULE` is that `T::ULE` must be encoded in little-endian with 1-byte alignment. When accessing
+/// items from `ZeroVec<T>`, we fetch the `T::ULE`, convert it on the fly to `T`, and return `T` by
+/// value.
+///
+/// Benchmarks can be found in the project repository. We found that for common operations on small
+/// and large vectors, `ZeroVec<T>` performs from slightly faster to 15% slower than `Vec<T>`.
+/// However, the main performance improvement on `ZeroVec<T>` is when deserializing from a byte
+/// array; `ZeroVec<T>` deserializes 80% faster than `Vec<T>` in Serde Bincode, and it does not
+/// require any heap allocations.
+///
+/// # Safety
+///
+/// `ZeroVec<T>` contains no unsafe code. However, the conversion from `&[u8]` to `&[T::ULE]` may
+/// be unsafe. For more information, see the [`ule`] module.
+///
+/// # Example
+///
+/// ```
+/// use zerovec::ZeroVec;
+///
+/// // The little-endian bytes correspond to the numbers on the following line.
+/// let bytes: &[u8] = &[0xD3, 0x00, 0x19, 0x01, 0xA5, 0x01, 0xCD, 0x01];
+/// let nums: &[u16] = &[211, 281, 421, 461];
+///
+/// // Conversion from &[u8] to &[u16::ULE] is infallible.
+/// let zerovec: ZeroVec<u16> = ZeroVec::try_from_bytes(bytes).expect("infallible");
+///
+/// assert!(matches!(zerovec, ZeroVec::Borrowed(_)));
+/// assert_eq!(zerovec.get(2), Some(421));
+/// assert_eq!(zerovec, nums);
+/// ```
+///
+/// [`ule`]: crate::ule
+#[non_exhaustive]
+#[derive(Clone)]
+pub enum ZeroVec<'a, T>
+where
+    T: AsULE + ?Sized,
+{
+    Owned(Vec<T::ULE>),
+    Borrowed(&'a [T::ULE]),
```

Comment:
idea (nb): Perhaps we should have this wrapped in a separate type much like `VarZeroVec` so that we can expose it, and people who do not need Cow-like behavior can use it? This could be a followup.

Reply:
Yeah I was thinking that might be a good idea.  I'd like to land this first though and work on that in a follow-up.  https://github.com/unicode-org/icu4x/issues/652